bolt/metrics.go

```diff
@@ -84,38 +85,92 @@ func (c *Client) Describe(ch chan<- *prometheus.Desc) {
 	ch <- dashboardsDesc
 	ch <- scrapersDesc
 	ch <- telegrafsDesc
-	ch <- telegrafPluginsDesc
 	ch <- boltWritesDesc
 	ch <- boltReadsDesc
+
+	c.pluginsCollector.Describe(ch)
 }
 
-type instaTicker struct {
-	tick   chan struct{}
-	timeCh <-chan time.Time
+type pluginMetricsCollector struct {
+	ticker     *time.Ticker
+	tickerDone chan struct{}
+	cache      map[string]float64
+	cacheMu    sync.RWMutex
 }
 
-var (
-	// ticker is this influx' timer for when to renew the cache of configured plugin metrics.
-	ticker *instaTicker
-	// telegrafPlugins is a cache of this influx' metrics of configured plugins.
-	telegrafPlugins = map[string]float64{}
-)
+func (c *pluginMetricsCollector) Open(db *bolt.DB) {
+	go c.pollTelegrafStats(db)
+}
 
-// Initialize a simple channel that will instantly "tick",
-// backed by a time.Ticker's channel.
-func init() {
-	ticker = &instaTicker{
-		tick:   make(chan struct{}, 1),
-		timeCh: time.NewTicker(time.Minute * 59).C,
+func (c *pluginMetricsCollector) pollTelegrafStats(db *bolt.DB) {
+	for {
+		select {
+		case <-c.tickerDone:
+			return
+		case <-c.ticker.C:
+			c.refreshTelegrafStats(db)
+		}
 	}
+}
+
+func (c *pluginMetricsCollector) refreshTelegrafStats(db *bolt.DB) {
+	c.cacheMu.Lock()
+	defer c.cacheMu.Unlock()
+
+	// Clear plugins from last check.
+	c.cache = map[string]float64{}
+
+	// Loop through all registered plugins.
+	_ = db.View(func(tx *bolt.Tx) error {
+		rawPlugins := [][]byte{}
+		_ = tx.Bucket(telegrafPluginsBucket).ForEach(func(k, v []byte) error {
+			rawPlugins = append(rawPlugins, v)
+			return nil
+		})
 
-	ticker.tick <- struct{}{}
+		for _, v := range rawPlugins {
+			pStats := map[string]float64{}
+			if err := json.Unmarshal(v, &pStats); err != nil {
+				return err
+			}
 
-	go func() {
-		for range ticker.timeCh {
-			ticker.tick <- struct{}{}
+			for k, v := range pStats {
+				c.cache[k] += v
+			}
 		}
-	}()
+
+		return nil
+	})
+}
+
+func (c *pluginMetricsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- telegrafPluginsDesc
+}
+
+func (c *pluginMetricsCollector) Collect(ch chan<- prometheus.Metric) {
+	c.cacheMu.RLock()
+	defer c.cacheMu.RUnlock()
+
+	for k, v := range c.cache {
+		ch <- prometheus.MustNewConstMetric(
+			telegrafPluginsDesc,
+			prometheus.GaugeValue,
+			v,
+			k, // Adds a label for plugin type.name.
+		)
+	}
+}
+
+func (c *pluginMetricsCollector) Close() {
+	close(c.tickerDone)
```

Comment:
Do you care if `refreshTelegrafStats` was started earlier and is still running even after you call Close?

Reply:
It probably doesn't matter since we're ignoring errors within `refreshTelegrafStats`, but I'll throw a `c.cacheMu.Lock()` at the front of this to be safe